include/color_spinor.h

```diff
@@ -919,6 +919,28 @@ namespace quda {
     return dot;
   }
 
+  /**
+     @brief Compute the color contraction over color at spin s
+     dot = \sum_s,c a(s,c) * b(s,c)
+     @param a Left-hand side ColorSpinor
+     @param b Right-hand side ColorSpinor
+     @return The color contraction
+  */
+  template <typename Float, int Nc, int Ns>
+  __device__ __host__ inline complex<Float> colorContract(const ColorSpinor<Float, Nc, Ns> &a,
+							  const ColorSpinor<Float, Nc, Ns> &b, int sa, int sb)
+  {
+    complex<Float> dot = 0;
+    for (int c = 0; c < Nc; c++) {
+      dot.x += a(sa, c).real() * b(sb, c).real();
+      dot.x -= a(sa, c).imag() * b(sb, c).imag();
+      dot.y += a(sa, c).real() * b(sb, c).imag();
+      dot.y += a(sa, c).imag() * b(sb, c).real();
+    }
```

Comment:
The correct way to write terms like this with modern C++ complex is something like the following.
```c++
      dot.real(dot.real() + a(sa, c).real() * b(sb, c).real());
      dot.real(dot.real() - a(sa, c).imag() * b(sb, c).imag());
      dot.imag(dot.imag() + a(sa, c).real() * b(sb, c).imag());
      dot.imag(dot.imag() + a(sa, c).imag() * b(sb, c).real());
```
It looks a bit messy, but probably good if we move away from the `.x` and `.y` in favour of more standard C++.

Reply:
I'll change the other two instances that use `dot.x` and `dot.y` while I'm at it.